server/mqtt.go

```diff
@@ -4476,33 +4570,43 @@ func (sess *mqttSession) ensurePubRelConsumerSubscription(c *client) error {
 		maxAckPending = mqttDefaultMaxAckPending
 	}
 
+	sess.mu.Lock()
+	pubRelSubscribed := sess.pubRelSubscribed
+	pubRelSubject := sess.pubRelSubject
+	pubRelDeliverySubjectB := sess.pubRelDeliverySubjectB
+	pubRelDeliverySubject := sess.pubRelDeliverySubject
+	pubRelConsumer := sess.pubRelConsumer
+	tmaxack := sess.tmaxack
+	idHash := sess.idHash
+	sess.mu.Unlock()
+
 	// Subscribe before the consumer is created so we don't loose any messages.
-	if !sess.pubRelSubscribed {
-		_, err := c.processSub(sess.pubRelDeliverySubjectB, nil, sess.pubRelDeliverySubjectB,
+	if !pubRelSubscribed {
+		_, err := c.processSub(pubRelDeliverySubjectB, nil, pubRelDeliverySubjectB,
```

Comment:
You are not calling this under the session lock. I know it is important for the other subs, so I assume that you understand that it is not needed for `mqttDeliverPubRelCb`.

Reply:
Right. This is neither an "mqtt" sub (doesn't need to be initialized as such), nor are there sync issues with other messages.

The main reason I took out the overarching session lock was because of creating the consumer in the step later. That one MUST be out of a lock since there may be a sub listening on `#` and such.

Reply:
Correct. Call to processSub() could be again, if there is no need (synchronization with callback, etc..), then please don't. Hence my comment about the comment at the top of the function stating that we lock for the entire duration that seem wrong...

Reply:
Sorry, took the comment out, it was wrong/leftover from the past.